worker/rhpv3.go

```diff
@@ -837,6 +837,191 @@ func RPCReadRegistry(t *rhpv3.Transport, payment rhpv3.PaymentMethod, key rhpv3.
+	}
+
+	// Perform gouging checks.
+	if gc := GougingCheckerFromContext(ctx).Check(nil, &pt); gc.Gouging() {
+		return rhpv2.ContractRevision{}, nil, fmt.Errorf("host gouging during renew: %v", gc.Reasons())
+	}
+
+	// Prepare the signed transaction that contains the final revision as well
+	// as the new contract
+	wprr, err := w.bus.WalletPrepareRenew(ctx, *rev, rrr.HostAddress, rrr.RenterAddress, renterKey, rrr.RenterFunds, rrr.NewCollateral, rrr.HostKey, pt, rrr.EndHeight, rrr.WindowSize)
+	if err != nil {
+		return rhpv2.ContractRevision{}, nil, err
+	}
+
+	// Starting from here, we need to make sure to release the txn on error.
+	defer func() {
+		if err != nil {
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			discardErr := w.bus.WalletDiscard(ctx, wprr.TransactionSet[len(wprr.TransactionSet)-1])
+			if discardErr != nil {
+				w.logger.Errorf("failed to discard txn after failed renewal: %v", discardErr)
+			}
+		}
+	}()
+
+	txnSet := wprr.TransactionSet
+	parents, txn := txnSet[:len(txnSet)-1], txnSet[len(txnSet)-1]
+
+	// Sign only the revision and contract. We can't sign everything because
+	// then the host can't add its own outputs.
+	h := types.NewHasher()
+	txn.FileContracts[0].EncodeTo(h.E)
+	txn.FileContractRevisions[0].EncodeTo(h.E)
+	finalRevisionSignature := renterKey.SignHash(h.Sum())
+
+	// Send the request.
+	req := rhpv3.RPCRenewContractRequest{
+		TransactionSet:         txnSet,
+		RenterKey:              rev.UnlockConditions.PublicKeys[0],
+		FinalRevisionSignature: finalRevisionSignature,
+	}
+	if err = s.WriteResponse(&req); err != nil {
+		return rhpv2.ContractRevision{}, nil, err
+	}
+
+	// Incorporate the host's additions.
+	var hostAdditions rhpv3.RPCRenewContractHostAdditions
+	if err = s.ReadResponse(&hostAdditions, 4096); err != nil {
+		return rhpv2.ContractRevision{}, nil, err
+	}
+	parents = append(parents, hostAdditions.Parents...)
+	txn.SiacoinInputs = append(txn.SiacoinInputs, hostAdditions.SiacoinInputs...)
+	txn.SiacoinOutputs = append(txn.SiacoinOutputs, hostAdditions.SiacoinOutputs...)
+	finalRevRenterSig := types.TransactionSignature{
+		ParentID:       types.Hash256(rev.ParentID),
+		PublicKeyIndex: 0, // renter key is first
+		CoveredFields: types.CoveredFields{
+			FileContracts:         []uint64{0},
+			FileContractRevisions: []uint64{0},
+		},
+		Signature: finalRevisionSignature[:],
+	}
+	finalRevHostSig := types.TransactionSignature{
+		ParentID:       types.Hash256(rev.ParentID),
+		PublicKeyIndex: 1,
+		CoveredFields: types.CoveredFields{
+			FileContracts:         []uint64{0},
+			FileContractRevisions: []uint64{0},
+		},
+		Signature: hostAdditions.FinalRevisionSignature[:],
+	}
+	txn.Signatures = []types.TransactionSignature{finalRevRenterSig, finalRevHostSig}
+
+	// Sign the inputs we funded the txn with and cover the whole txn including
+	// the existing signatures.
+	cf := types.CoveredFields{
+		WholeTransaction: true,
```

Comment:
Wanted to make sure this is def. fine right, because I think we're coming from `ExplicitCoveredFields`

Reply:
The host will accept `ExplicitCoveredFields` too but I think in this case (since we explicitly set the `Signatures` field) it's almost identical. Except that `WholeTransaction` also covers the parent id of the signature, the public key index and the time lock which might be a bit more secure.